src/lib.rs

```diff
@@ -1195,26 +1195,71 @@ where
             StreamResult::Finished => Ok(StreamResult::Finished),
         }
     }
+}
 
-    pub async fn multi_valued_next<T: for<'b> serde::Deserialize<'b>>(
-        &mut self,
-    ) -> Result<StreamResult<T>, Error> {
-        let entry = self.next_inner().await?;
-        match entry {
-            StreamResult::Record(entry) => {
-                let json = LdapClient::create_json_multi_value(entry).unwrap();
-                let data = LdapClient::map_to_struct::<T>(json);
-                if let Err(err) = data {
-                    return Err(Error::Mapping(format!("Error mapping record: {:?}", err)));
+impl<'a, S, A> futures::stream::Stream for Stream<'a, S, A>
+where
+    S: AsRef<str> + Send + Sync + 'a,
+    A: AsRef<[S]> + Send + Sync + 'a,
+{
+    type Item = Result<Record, Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Record, Error>>> {
+        let poll = self.next_inner().boxed().as_mut().poll(cx);
+        match poll {
+            Poll::Ready(result) => match result {
+                Ok(result) => match result {
+                    StreamResult::Record(record) => Poll::Ready(Some(Ok(Record {
+                        search_entry: record,
+                    }))),
+                    StreamResult::Done => Poll::Ready(None),
+                    StreamResult::Finished => Poll::Ready(None),
+                },
+                Err(er) => {
+                    return Poll::Ready(Some(Err(er)));
                 }
-                return Ok(StreamResult::Record(data.unwrap()));
+            },
+            Poll::Pending => {
+                if self.count == self.limit {
+                    return Poll::Ready(None);
+                }
+                // Ensure the task is woken when the next record is ready
+                cx.waker().wake_by_ref();
+                Poll::Pending
             }
-            StreamResult::Done => Ok(StreamResult::Done),
-            StreamResult::Finished => Ok(StreamResult::Finished),
         }
     }
 }
 
+pub struct Record {
+    search_entry: SearchEntry,
+}
+
+impl Record {
+    pub fn to_record<T: for<'b> serde::Deserialize<'b>>(self) -> Result<T, Error> {
+        let json = LdapClient::create_json_signle_value(self.search_entry).unwrap();
```

Comment:
Not related to streams anymore but this intermediate conversion to JSON seems unnecessary? `serde_json` could probably be refactored out completely, but perhaps that's a matter for another PR.

Reply:
Yes, this is something that I want to improve. I will work on this in a separate PR.